include/ccf/pal/attestation.h

```diff
@@ -0,0 +1,373 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#pragma once
+
+#include "ccf/ds/logger.h"
+#include "ccf/ds/quote_info.h"
+
+#if !defined(INSIDE_ENCLAVE) || defined(VIRTUAL_ENCLAVE)
+#  include "ccf/crypto/pem.h"
+#  include "ccf/crypto/verifier.h"
+#  include "ccf/pal/attestation_sev_snp.h"
+#  include "clients/rpc_tls_client.h"
+#  include "crypto/ecdsa.h"
+
+#  include <fcntl.h>
+#  include <sys/ioctl.h>
+#  include <unistd.h>
+#else
+#  include "ccf/pal/attestation_sgx.h"
+
+#  include <openenclave/attestation/attester.h>
+#endif
+
+namespace ccf::pal
+{
+#if !defined(INSIDE_ENCLAVE) || defined(VIRTUAL_ENCLAVE)
+
+  static QuoteInfo generate_quote(attestation_report_data&& report_data)
+  {
+    QuoteInfo node_quote_info = {};
+    auto is_sev_snp = access(snp::DEVICE, F_OK) == 0;
+
+    // If there is no SEV-SNP device, assume we are using insecure virtual
+    // quotes
+    if (!is_sev_snp)
+    {
+      node_quote_info.format = QuoteFormat::insecure_virtual;
+    }
+
+    else
+    {
+      node_quote_info.format = QuoteFormat::amd_sev_snp_v1;
+      int fd = open(snp::DEVICE, O_RDWR | O_CLOEXEC);
+      if (fd < 0)
+      {
+        throw std::logic_error("Failed to open \"/dev/sev\"");
+      }
+
+      snp::AttestationReq req = {};
+      snp::AttestationResp resp = {};
+
+      // Arbitrary report data
+      memcpy(req.report_data, report_data.data(), attestation_report_data_size);
+
+      snp::GuestRequest payload = {
+        .req_msg_type = snp::MSG_REPORT_REQ,
+        .rsp_msg_type = snp::MSG_REPORT_RSP,
+        .msg_version = 1,
+        .request_len = sizeof(req),
+        .request_uaddr = (uint64_t)(void*)&req,
+        .response_len = sizeof(resp),
+        .response_uaddr = (uint64_t)(void*)&resp,
+        .error = 0};
+
+      int rc = ioctl(fd, SEV_SNP_GUEST_MSG_REPORT, &payload);
+      if (rc < 0)
+      {
+        CCF_APP_FAIL("IOCTL call failed: {}", strerror(errno));
+        CCF_APP_FAIL("Payload error: {}", payload.error);
+        throw std::logic_error(
+          "Failed to issue ioctl SEV_SNP_GUEST_MSG_REPORT");
+      }
+
+      auto quote = &resp.report;
+      auto quote_bytes = reinterpret_cast<uint8_t*>(&resp.report);
+      node_quote_info.quote.assign(quote_bytes, quote_bytes + resp.report_size);
+
+      client::RpcTlsClient client{
+        "americas.test.acccache.azure.net", // TODO: Make Configurable
```

Comment:
Needs a ticket

Reply:
Default to AMD

Reply:
Looks like using the AMD URL returns a certificate that when parsed as PEM is slightly different to the one from Azure. I don't know if there's an issue with our/openssl's parsing or if the certificate is genuinely different, but for the sake of getting something that works merged, I propose we add the AMD server as default as part of https://github.com/microsoft/CCF/issues/4144

Reply:
Sounds good to me.